### ISSUE TITLE
mzIdentML file malformed

### ISSUE THREAD

**Issue by kaipot** [crux-toolkit/crux-toolkit#423]
> Dear team,
> 
> I am using Crux Percolator to process MS-GF+ output converted to PIN as per https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3975676/
> 
> I am using Crux Percolator rather than the standard Percolator because I need MzIdentML files for viewing in Pride Inspector.
> 
> However, the file crashes Pride Inspector and the mzIdentML validator. When I use a general purpose XML validator, I hit a range of issues:
>  - the schema reference doesn't work - possibly out of date
>  - when updated, there is a missing cvList
>  - when I place a generic cvList, I then receive a litany of errors regarding missing attributes (e.g. searchDatabase_ref from DBSequences and spectraData_ref from SpectrumIdentificationResults, and DBSequence IDs fail to refer to expected keys)
> 
> I am unsure whether this is because I have made a mistake in running crux percolator, or if the mzIdentML does not follow the schema well.
> 
> I would be very grateful for your advice.
> 
> Best wishes,
> 
> Andrew
> Clinical Research Fellow
> Imperial College London

**Comment by ypriverol** [comment]
Dear @percolator @MatthewThe @grobot @CharlesEGrant @kaipot: 

I'm Yasset Lead Dev of PRIDE team and more than one user has pointed out this issue. We can't do anything about that. Please try to give some priority to this in order to support Complete submissions from Tide/Crux/Percolator. 

Regards 
Yasset

**Comment by kaipot** [comment]
I've just committed a change that will add the cvList and it should be deployed to crux.ms for download soon.

**Comment by ypriverol** [comment]
Thanks a lot, let me know when is ready in order to test it. I think other sections imporant for mzIdentML were also missing. Such as: 

- Identification Protocols. 

If you want you can give me an example and I will test it here. 

REgards 
Yasset

**Comment by wsnoble** [comment]
This should be on the main page now. Can you test it Yasset and let us know if the problem has been addressed?

**Comment by ypriverol** [comment]
@wsnoble the version you available in the web 3.2 includes the new changes?

**Comment by wsnoble** [comment]
Yes, we use a continuous build system, and I'm told that Kaipo's changes
have been automatically propagated to the crux.ms page.

**Comment by CharlesEGrant** [comment]
Be sure to use the button labeled "I agree to the licensing terms, download the most recent build of Crux".

**Comment by ypriverol** [comment]
Yes, thanks I notice that. Testing locally now. I will keep you posted about my results.

**Comment by ypriverol** [comment]
Dear @kaipot still the file is invalid: 

Major errors: 

- [ ] DBSequence: Schemas validity error : Element '{http://psidev.info/psi/pi/mzIdentML/1.1}DBSequence': The attribute 'searchDatabase_ref' is required but missing.

- [ ] SpectrumIdentificationProtocol: Schemas validity error : Element '{http://psidev.info/psi/pi/mzIdentML/1.1}SpectrumIdentificationProtocol': The attribute 'analysisSoftware_ref' is required but missing.

- [ ]  SpectraData: Schemas validity error : Element '{http://psidev.info/psi/pi/mzIdentML/1.1}SpectraData': Missing child element(s). Expected is one of ( {http://psidev.info/psi/pi/mzIdentML/1.1}ExternalFormatDocumentation, {http://psidev.info/psi/pi/mzIdentML/1.1}FileFormat, {http://psidev.info/psi/pi/mzIdentML/1.1}SpectrumIDFormat ).

I'm using the following command to validate the schema: 

> xmllint --noout --valid --schema /Users/yperez/IdeaProjects/github-repo/utilitis-organization/mzIdentML/src/main/resources/mzIdentML1.1.0.xsd /Users/yperez/work/ms_work/nextflow/nf-workflows/work/e9/41184813ba5a090457a1960e655be9/crux-output/tide-search.mzid

I can provide you a couple of examples mzIdentML from other softwares to have a look if you want.

**Comment by wsnoble** [comment]
Sorry for the slow response here. Can you send the mzIdentML file from Crux that you are attempting to validate, and the command line that generated it?

**Comment by ypriverol** [comment]
Professor, 

I don't think the errors I got are case specific. You can take any analysis in your side export mzIdentml and run the previous command I gave you and you will see the errors. This is the command I use to validate the tools: 

xmllint --noout --valid --schema 
mzIdentML1.1.0.xsd  file-output.mzid

**Comment by wsnoble** [comment]
For example, this one:

https://github.com/crux-toolkit/crux-toolkit/blob/master/test/smoke-tests/good_results/psmconv-from-txt2.mzid

**Comment by ypriverol** [comment]
Yes, if you try to validate that file it produces the same errors.

**Comment by CharlesEGrant** [comment]
If xmllint is used to validate Crux's mzid output using for example

xmllint --noout --schema mzIdentML1.1.0.xsd psmconv-from-txt1.mzid &> schema-errors.txt

For each sequence we get the following error:

psmconv-from-txt1.mzid:9: element DBSequence: Schemas validity error : Element '{http://psidev.info/psi/pi/mzIdentML/1.1}DBSequence': The attribute 'searchDatabase_ref' is required but missing. 
Indeed, looking at the schema, 'searchDatabase_ref is a required attribute of a DBSequence tag, and we aren't writing it out in Crux when we generate mzid output.

Looking at the pwiz code pwiz/data/identdata/TextWriter.hpp it looks like the value of that attribute is taken from the id member of type SearchDatabase. We currently aren't setting up a SearchDatabase object for our DBSequences. In fact, there is a comment at line 195 of src/io/MzIdentMLWriter.cpp that we may not have the protein database and this may be a problem. It may be that we just have to create a dummy SearchDatabase instance, just to meet the requirements of the schema.

A useful example of creating a SearchDatabase object is found in pwiz/data/identdat/examples.cpp at line 189.
